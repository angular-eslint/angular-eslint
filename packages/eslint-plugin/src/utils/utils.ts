import type { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';

export const objectKeys = Object.keys as <T>(
  o: T,
) => ReadonlyArray<Extract<keyof T, string>>;

export enum AngularClassDecorators {
  Component = 'Component',
  Directive = 'Directive',
  Injectable = 'Injectable',
  NgModule = 'NgModule',
  Pipe = 'Pipe',
}

enum AngularConstructorParameterDecorators {
  Attribute = 'Attribute',
  Host = 'Host',
  Inject = 'Inject',
  Optional = 'Optional',
  Self = 'Self',
  SkipSelf = 'SkipSelf',
}

enum AngularMethodDecorators {
  HostListener = 'HostListener',
}

enum AngularPropertyAccessorDecorators {
  ContentChild = 'ContentChild',
  ContentChildren = 'ContentChildren',
  HostBinding = 'HostBinding',
  Input = 'Input',
  Output = 'Output',
  ViewChild = 'ViewChild',
  ViewChildren = 'ViewChildren',
}

export const AngularInnerClassDecorators = {
  ...AngularConstructorParameterDecorators,
  ...AngularMethodDecorators,
  ...AngularPropertyAccessorDecorators,
};

export enum AngularLifecycleInterfaces {
  AfterContentChecked = 'AfterContentChecked',
  AfterContentInit = 'AfterContentInit',
  AfterViewChecked = 'AfterViewChecked',
  AfterViewInit = 'AfterViewInit',
  DoBootstrap = 'DoBootstrap',
  DoCheck = 'DoCheck',
  OnChanges = 'OnChanges',
  OnDestroy = 'OnDestroy',
  OnInit = 'OnInit',
}

export enum AngularLifecycleMethods {
  ngAfterContentChecked = 'ngAfterContentChecked',
  ngAfterContentInit = 'ngAfterContentInit',
  ngAfterViewChecked = 'ngAfterViewChecked',
  ngAfterViewInit = 'ngAfterViewInit',
  ngDoBootstrap = 'ngDoBootstrap',
  ngDoCheck = 'ngDoCheck',
  ngOnChanges = 'ngOnChanges',
  ngOnDestroy = 'ngOnDestroy',
  ngOnInit = 'ngOnInit',
}

export const OPTION_STYLE_CAMEL_CASE = 'camelCase';
export const OPTION_STYLE_KEBAB_CASE = 'kebab-case';
export type SelectorStyle =
  | typeof OPTION_STYLE_CAMEL_CASE
  | typeof OPTION_STYLE_KEBAB_CASE;

export type AngularClassDecoratorKeys = keyof typeof AngularClassDecorators;
export type AngularInnerClassDecoratorKeys = Exclude<
  keyof typeof AngularInnerClassDecorators,
  number
>;
export type AngularLifecycleInterfaceKeys = keyof typeof AngularLifecycleInterfaces;
export type AngularLifecycleMethodKeys = keyof typeof AngularLifecycleMethods;

export const angularClassDecoratorKeys = objectKeys(AngularClassDecorators);
export const angularInnerClassDecoratorKeys = objectKeys(
  AngularInnerClassDecorators,
);
export const angularLifecycleInterfaceKeys = objectKeys(
  AngularLifecycleInterfaces,
);
export const angularLifecycleMethodKeys = objectKeys(AngularLifecycleMethods);

export const ANGULAR_CLASS_DECORATOR_LIFECYCLE_METHOD_MAPPER: ReadonlyMap<
  AngularClassDecoratorKeys,
  ReadonlySet<AngularLifecycleMethodKeys>
> = new Map([
  [
    AngularClassDecorators.Component,
    new Set([
      AngularLifecycleMethods.ngAfterContentChecked,
      AngularLifecycleMethods.ngAfterContentInit,
      AngularLifecycleMethods.ngAfterViewChecked,
      AngularLifecycleMethods.ngAfterViewInit,
      AngularLifecycleMethods.ngOnChanges,
      AngularLifecycleMethods.ngOnDestroy,
      AngularLifecycleMethods.ngOnInit,
      AngularLifecycleMethods.ngDoCheck,
    ]),
  ],
  [
    AngularClassDecorators.Directive,
    new Set([
      AngularLifecycleMethods.ngAfterContentChecked,
      AngularLifecycleMethods.ngAfterContentInit,
      AngularLifecycleMethods.ngAfterViewChecked,
      AngularLifecycleMethods.ngAfterViewInit,
      AngularLifecycleMethods.ngOnChanges,
      AngularLifecycleMethods.ngOnDestroy,
      AngularLifecycleMethods.ngOnInit,
      AngularLifecycleMethods.ngDoCheck,
    ]),
  ],
  [
    AngularClassDecorators.Injectable,
    new Set<AngularLifecycleMethodKeys>([AngularLifecycleMethods.ngOnDestroy]),
  ],
  [
    AngularClassDecorators.NgModule,
    new Set<AngularLifecycleMethodKeys>([
      AngularLifecycleMethods.ngDoBootstrap,
    ]),
  ],
  [
    AngularClassDecorators.Pipe,
    new Set<AngularLifecycleMethodKeys>([AngularLifecycleMethods.ngOnDestroy]),
  ],
]);

export const ANGULAR_INNER_CLASS_DECORATORS: ReadonlySet<AngularInnerClassDecoratorKeys> = new Set(
  angularInnerClassDecoratorKeys,
);

export const ANGULAR_CLASS_DECORATORS: ReadonlySet<AngularClassDecoratorKeys> = new Set(
  angularClassDecoratorKeys,
);

export const ANGULAR_CLASS_DECORATOR_MAPPER: ReadonlyMap<
  AngularClassDecoratorKeys,
  ReadonlySet<AngularInnerClassDecoratorKeys>
> = new Map([
  [AngularClassDecorators.Component, ANGULAR_INNER_CLASS_DECORATORS],
  [AngularClassDecorators.Directive, ANGULAR_INNER_CLASS_DECORATORS],
  [
    AngularClassDecorators.Injectable,
    new Set([
      AngularInnerClassDecorators.Host,
      AngularInnerClassDecorators.Inject,
      AngularInnerClassDecorators.Optional,
      AngularInnerClassDecorators.Self,
      AngularInnerClassDecorators.SkipSelf,
    ]),
  ],
  [
    AngularClassDecorators.NgModule,
    new Set([
      AngularInnerClassDecorators.Host,
      AngularInnerClassDecorators.Inject,
      AngularInnerClassDecorators.Optional,
      AngularInnerClassDecorators.Self,
      AngularInnerClassDecorators.SkipSelf,
    ]),
  ],
  [
    AngularClassDecorators.Pipe,
    new Set([
      AngularInnerClassDecorators.Host,
      AngularInnerClassDecorators.Inject,
      AngularInnerClassDecorators.Optional,
      AngularInnerClassDecorators.Self,
      AngularInnerClassDecorators.SkipSelf,
    ]),
  ],
]);

/**
 * SECTION START:
 * Equivalents of utils exported by TypeScript itself for its own AST
 */

export function isCallExpression(
  node: TSESTree.Node,
): node is TSESTree.CallExpression {
  return node.type === 'CallExpression';
}

export function isIdentifier(node: TSESTree.Node): node is TSESTree.Identifier {
  return node.type === 'Identifier';
}

export function isMemberExpression(
  node: TSESTree.Node,
): node is TSESTree.MemberExpression {
  return node.type === 'MemberExpression';
}

export function isClassDeclaration(
  node: TSESTree.Node,
): node is TSESTree.ClassDeclaration {
  return node.type === 'ClassDeclaration';
}

export function isObjectExpression(
  node: TSESTree.Node,
): node is TSESTree.ObjectExpression {
  return node.type === 'ObjectExpression';
}

export function isArrayExpression(
  node: TSESTree.Node,
): node is TSESTree.ArrayExpression {
  return node.type === 'ArrayExpression';
}

export function isProperty(node: TSESTree.Node): node is TSESTree.Property {
  return node.type === 'Property';
}

function isProgram(node: TSESTree.Node): node is TSESTree.Program {
  return node.type === 'Program';
}

export function isLiteral(node: TSESTree.Node): node is TSESTree.Literal {
  return node.type === 'Literal';
}

export function isTemplateLiteral(
  node: TSESTree.Node,
): node is TSESTree.TemplateLiteral {
  return node.type === 'TemplateLiteral';
}

export function isImportDeclaration(
  node: TSESTree.Node,
): node is TSESTree.ImportDeclaration {
  return node.type === 'ImportDeclaration';
}

function isImportSpecifier(
  node: TSESTree.Node,
): node is TSESTree.ImportSpecifier {
  return node.type === 'ImportSpecifier';
}

type LiteralWithStringValue = TSESTree.Literal & {
  value: string;
};

/**
 * ESTree does not differentiate between different types of Literals at the AST level,
 * but it is a common thing to need to do, so this utility and interface are here to
 * avoid repeated typeof checks on the node's value.
 */
export function isLiteralWithStringValue(
  node: TSESTree.Node,
): node is LiteralWithStringValue {
  return node.type === 'Literal' && typeof node.value === 'string';
}

export function isMethodDefinition(
  node: TSESTree.Node,
): node is TSESTree.MethodDefinition {
  return node.type === 'MethodDefinition';
}

export function isSuper(node: TSESTree.Node): node is TSESTree.Super {
  return node.type === 'Super';
}

/**
 * SECTION END:
 * Equivalents of utils exported by TypeScript itself for its own AST
 */

export function getNearestNodeFrom<T extends TSESTree.Node>(
  { parent }: TSESTree.Node,
  predicate: (parent: TSESTree.Node) => parent is T,
): T | null {
  while (parent && !isProgram(parent)) {
    if (predicate(parent)) {
      return (parent as unknown) as T;
    }

    parent = parent.parent as TSESTree.Node | undefined;
  }

  return null;
}

export function getImportDeclarations(
  node: TSESTree.Node,
  moduleName: string,
): readonly TSESTree.ImportDeclaration[] | undefined {
  let parentNode: TSESTree.Node | undefined = node;

  while ((parentNode = parentNode.parent)) {
    if (parentNode.type !== 'Program') continue;

    return parentNode.body.filter(
      (node): node is TSESTree.ImportDeclaration =>
        isImportDeclaration(node) && node.source.value === moduleName,
    );
  }

  return parentNode;
}

export function getImplementsRemoveFix(
  sourceCode: Readonly<TSESLint.SourceCode>,
  classDeclaration: TSESTree.ClassDeclaration,
  interfaceName: string,
  fixer: TSESLint.RuleFixer,
): TSESLint.RuleFix | TSESLint.RuleFix[] {
  const { implements: classImplements } = classDeclaration;

  if (!classImplements) return [];

  const identifier = classImplements
    .map(({ expression }) => expression)
    .filter(isIdentifier)
    .find(({ name }) => name === interfaceName);

  if (!identifier) return [];

  const isFirstInterface = classImplements[0].expression === identifier;
  const isLastInterface =
    classImplements[classImplements.length - 1].expression === identifier;
  const isSingleInterface = isFirstInterface && isLastInterface;

  if (isSingleInterface && classDeclaration.id) {
    return fixer.removeRange([
      classDeclaration.id.range[1],
      classImplements[0].range[1],
    ]);
  }

  const tokenAfterInterface = sourceCode.getTokenAfter(identifier);

  if (isFirstInterface && tokenAfterInterface) {
    return fixer.removeRange([
      identifier.range[0],
      tokenAfterInterface.range[1],
    ]);
  }

  const tokenBeforeInterface = sourceCode.getTokenBefore(identifier);

  if (tokenBeforeInterface) {
    return fixer.removeRange([
      tokenBeforeInterface.range[0],
      identifier.range[1],
    ]);
  }

  return [];
}

function getImportDeclarationSpecifier(
  importDeclarations: readonly TSESTree.ImportDeclaration[],
  importedName: string,
) {
  for (const importDeclaration of importDeclarations) {
    const importSpecifier = importDeclaration.specifiers.find(
      (importClause): importClause is TSESTree.ImportSpecifier => {
        return (
          isImportSpecifier(importClause) &&
          importClause.imported.name === importedName
        );
      },
    );

    if (importSpecifier) {
      return { importDeclaration, importSpecifier } as const;
    }
  }

  return undefined;
}

export function getImportRemoveFix(
  sourceCode: Readonly<TSESLint.SourceCode>,
  importDeclarations: readonly TSESTree.ImportDeclaration[],
  importedName: string,
  fixer: TSESLint.RuleFixer,
): TSESLint.RuleFix | TSESLint.RuleFix[] {
  const { importDeclaration, importSpecifier } =
    getImportDeclarationSpecifier(importDeclarations, importedName) ?? {};

  if (!importDeclaration || !importSpecifier) return [];

  const isFirstImportSpecifier =
    importDeclaration.specifiers[0] === importSpecifier;
  const isLastImportSpecifier =
    importDeclaration.specifiers[importDeclaration.specifiers.length - 1] ===
    importSpecifier;
  const isSingleImportSpecifier =
    isFirstImportSpecifier && isLastImportSpecifier;

  if (isSingleImportSpecifier) {
    return fixer.remove(importDeclaration);
  }

  const tokenAfterImportSpecifier = sourceCode.getTokenAfter(importSpecifier);

  if (isFirstImportSpecifier && tokenAfterImportSpecifier) {
    return fixer.removeRange([
      importSpecifier.range[0],
      tokenAfterImportSpecifier.range[1],
    ]);
  }

  const tokenBeforeImportSpecifier = sourceCode.getTokenBefore(importSpecifier);

  if (tokenBeforeImportSpecifier) {
    return fixer.removeRange([
      tokenBeforeImportSpecifier.range[0],
      importSpecifier.range[1],
    ]);
  }

  return [];
}

export const getClassName = (node: TSESTree.Node): string | undefined => {
  if (isClassDeclaration(node)) {
    return node.id?.name;
  }
  if (node.parent) {
    return getClassName(node.parent);
  }
  return undefined;
};

export const getDecorator = (
  node: TSESTree.ClassDeclaration,
  decoratorName: string,
): TSESTree.Decorator | undefined => {
  if (!node.decorators) {
    return undefined;
  }
  return node.decorators.find(
    (decorator) =>
      isCallExpression(decorator.expression) &&
      decorator.expression.arguments &&
      decorator.expression.arguments.length > 0 &&
      getDecoratorName(decorator) === decoratorName,
  );
};

export const getAngularClassDecorator = ({
  decorators,
}: TSESTree.ClassDeclaration): AngularClassDecoratorKeys | undefined => {
  return decorators
    ?.map(getDecoratorName)
    .filter(isNotNullOrUndefined)
    .find(isAngularClassDecorator);
};

export const getDecoratorArgument = (
  decorator: TSESTree.Decorator,
): TSESTree.ObjectExpression | undefined => {
  const { expression } = decorator;
  if (
    !isCallExpression(expression) ||
    !expression.arguments ||
    expression.arguments.length === 0
  ) {
    return undefined;
  }
  const arg = expression.arguments[0];
  return isObjectExpression(arg) && arg.properties ? arg : undefined;
};

export const getDecoratorName = (
  decorator: TSESTree.Decorator,
): string | undefined => {
  const { expression } = decorator;

  if (isIdentifier(expression)) return expression.name;

  if (isCallExpression(expression) && isIdentifier(expression.callee)) {
    return expression.callee.name;
  }

  return undefined;
};

export const getPipeDecorator = (
  node: TSESTree.ClassDeclaration,
): TSESTree.Decorator | undefined => getDecorator(node, 'Pipe');

export const getSymbolName = (
  expression: TSESTree.TSClassImplements,
): string => {
  const { expression: childExpression } = expression;

  // TODO: Investigate "as any"
  return isMemberExpression(childExpression)
    ? (childExpression.property as any).name
    : (childExpression as any).name;
};

export const getDeclaredInterfaces = (
  node: TSESTree.ClassDeclaration,
): TSESTree.TSClassImplements[] => {
  return node.implements || [];
};

export const getDeclaredInterfaceNames = (
  node: TSESTree.ClassDeclaration,
): string[] => getDeclaredInterfaces(node).map(getSymbolName);

export const getDeclaredInterfaceName = (
  node: TSESTree.ClassDeclaration,
  value: string,
): string | undefined =>
  getDeclaredInterfaceNames(node).find(
    (interfaceName) => interfaceName === value,
  );

export const getDeclaredAngularLifecycleInterfaces = (
  node: TSESTree.ClassDeclaration,
): ReadonlyArray<AngularLifecycleInterfaceKeys> =>
  getDeclaredInterfaceNames(node).filter(
    isAngularLifecycleInterface,
  ) as ReadonlyArray<AngularLifecycleInterfaceKeys>;

export const getDeclaredAngularLifecycleMethods = (
  node: TSESTree.ClassDeclaration,
): ReadonlyArray<AngularLifecycleMethodKeys> =>
  getDeclaredMethods(node)
    .map(getMethodName)
    .filter(isAngularLifecycleMethod) as ReadonlyArray<
    AngularLifecycleMethodKeys
  >;

export const ANGULAR_LIFECYCLE_INTERFACES: ReadonlySet<AngularLifecycleInterfaceKeys> = new Set(
  angularLifecycleInterfaceKeys,
);

export const ANGULAR_LIFECYCLE_METHODS: ReadonlySet<AngularLifecycleMethodKeys> = new Set(
  angularLifecycleMethodKeys,
);

export const isAngularLifecycleInterface = (
  value: string,
): value is AngularLifecycleInterfaceKeys =>
  ANGULAR_LIFECYCLE_INTERFACES.has(value as AngularLifecycleInterfaceKeys);

export const isAngularLifecycleMethod = (
  value: string,
): value is AngularLifecycleMethodKeys =>
  ANGULAR_LIFECYCLE_METHODS.has(value as AngularLifecycleMethodKeys);

export const isAngularClassDecorator = (
  value: string,
): value is AngularClassDecoratorKeys =>
  ANGULAR_CLASS_DECORATORS.has(value as AngularClassDecoratorKeys);

export const isAngularInnerClassDecorator = (
  value: string,
): value is AngularInnerClassDecoratorKeys =>
  ANGULAR_INNER_CLASS_DECORATORS.has(value as AngularInnerClassDecoratorKeys);

/**
 * `ClassProperty` nodes can have different types of `key`s
 *
 * E.g.
 *
 * class Foo {
 *  a // Identifier
 * 'b' // Literal
 *  ['c'] // Literal
 * }
 */
export function getClassPropertyName(
  classProperty: TSESTree.ClassProperty,
): string {
  if (classProperty.key.type === 'Identifier') {
    return classProperty.key.name;
  }

  if (classProperty.key.type === 'Literal') {
    return classProperty.key.raw;
  }

  throw new Error(
    `Unexpected "ClassProperty.key.type" provided: ${classProperty.key.type}`,
  );
}

export const getDecoratorProperty = (
  decorator: TSESTree.Decorator,
  name: string,
): TSESTree.Property | undefined => {
  const arg = getDecoratorArgument(decorator);

  if (!arg || !isObjectExpression(arg)) return undefined;

  const properties = arg.properties as TSESTree.Property[];
  const property = properties.find(
    (prop) => prop.key && isIdentifier(prop.key) && prop.key.name === name,
  );

  if (!property || !isProperty(property)) return undefined;

  return property;
};

export const getDecoratorPropertyValue = (
  decorator: TSESTree.Decorator,
  name: string,
): TSESTree.Expression | TSESTree.Literal | undefined => {
  const property = getDecoratorProperty(decorator, name);
  if (!property) {
    return undefined;
  }

  /**
   * TODO: Investigate as any
   */
  return property.value as any;
};

export const getDeclaredMethods = (node: TSESTree.ClassDeclaration) => {
  return node.body.body.filter(isMethodDefinition);
};

export const getMethodName = (node: TSESTree.MethodDefinition): string => {
  if (isLiteral(node.key)) {
    return node.key.value as string;
  }
  return (node.key as TSESTree.Identifier).name;
};

export const getLifecycleInterfaceByMethodName = (
  methodName: AngularLifecycleMethodKeys,
): AngularLifecycleInterfaceKeys =>
  methodName.slice(2) as AngularLifecycleInterfaceKeys;

/**
 * Enforces the invariant that the input is an array.
 */
export function arrayify<T>(arg?: T | T[]): T[] {
  if (Array.isArray(arg)) {
    return arg;
  } else if (arg != undefined) {
    return [arg];
  } else {
    return [];
  }
}

// Needed because in the current Typescript version (TS 3.3.3333), Boolean() cannot be used to perform a null check.
// For more, see: https://github.com/Microsoft/TypeScript/issues/16655
export const isNotNullOrUndefined = <T>(
  input: null | undefined | T,
): input is T => input !== null && input !== undefined;

export const SelectorValidator = {
  attribute(selector: string): boolean {
    return selector.length !== 0;
  },

  camelCase(selector: string): boolean {
    return /^[a-zA-Z0-9[\]]+$/.test(selector);
  },

  element(selector: string): boolean {
    return selector !== null;
  },

  kebabCase(selector: string): boolean {
    return /^[a-z0-9-]+-[a-z0-9-]+$/.test(selector);
  },

  prefix(
    prefix: string,
    selectorStyle: SelectorStyle,
  ): (selector: string) => boolean {
    const regex = new RegExp(`^\\[?(${prefix})`);

    return (selector) => {
      if (!prefix) return true;

      if (!regex.test(selector)) return false;

      const suffix = selector.replace(regex, '');

      if (selectorStyle === OPTION_STYLE_CAMEL_CASE) {
        return !suffix || suffix[0] === suffix[0].toUpperCase();
      } else if (selectorStyle === OPTION_STYLE_KEBAB_CASE) {
        return !suffix || suffix[0] === '-';
      }

      throw Error('Invalid selector style!');
    };
  },
};

export const kebabToCamelCase = (value: string) =>
  value.replace(/-[a-zA-Z]/g, (x) => x[1].toUpperCase());

export function isImportedFrom(
  identifier: TSESTree.Identifier,
  module: string,
) {
  let parentNode: TSESTree.Node | undefined = identifier;
  while ((parentNode = parentNode.parent)) {
    if (parentNode.type !== 'Program') continue;
    return parentNode.body.some(
      (node) =>
        isImportDeclaration(node) &&
        (node.source as TSESTree.Literal).value === module &&
        node.specifiers.some(
          (specifier) =>
            isImportSpecifier(specifier) &&
            specifier.imported.name === identifier.name &&
            specifier.local.name === identifier.name,
        ),
    );
  }
  return false;
}

/**
 * Convert an array to human-readable text.
 */
export const toHumanReadableText = (items: readonly string[]): string => {
  const itemsLength = items.length;

  if (itemsLength === 1) {
    return `"${items[0]}"`;
  }

  return `${items
    .map((item) => `"${item}"`)
    .slice(0, itemsLength - 1)
    .join(', ')} or "${[...items].pop()}"`;
};

export const toPattern = (value: readonly unknown[]) =>
  RegExp(`^(${value.join('|')})$`);
