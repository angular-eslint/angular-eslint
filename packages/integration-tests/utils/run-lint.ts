/* eslint-disable @typescript-eslint/no-non-null-assertion */
import execa from 'execa';
import path from 'path';

import stripAnsi = require('strip-ansi');

const FIXTURES_DIR = path.join(__dirname, '../fixtures/');

function normalizeOutput(value: unknown): string {
  const ansiRemoved = stripAnsi(String(value));
  return (
    ansiRemoved
      .replace(
        new RegExp(`^${FIXTURES_DIR.replace(/\\/g, '\\\\')}(.*?)$`, 'gm'),
        (_, c1) => `__ROOT__/${c1.replace(/\\/g, '/')}`,
      )
      /**
       * This one is ridiculous...
       *
       * No matter what I try (including the use of `strip-ansi`), I do not seem to be
       * able to stop the inclusion of this extra `m` any other way (it is presumably
       * coming from an ANSI code becuase of its initial appearance in CI as [27m).
       * It happens locally and CI... but only exactly once.
       */
      .replace('mmax-len', 'max-len')
  );
}

export async function runLint(directory: string): Promise<string | undefined> {
  try {
    const subprocess = execa('npx', ['ng', 'lint'], {
      cwd: path.join(FIXTURES_DIR, directory),
    });

    /* eslint-disable @typescript-eslint/no-non-null-assertion */
    subprocess.stdout!.pipe(process.stdout);
    subprocess.stderr!.pipe(process.stderr);
    /* eslint-enable @typescript-eslint/no-non-null-assertion */

    const { stdout } = await subprocess;

    return normalizeOutput(stdout);
  } catch (error: any) {
    return normalizeOutput(error.stdout || error);
  }
}
