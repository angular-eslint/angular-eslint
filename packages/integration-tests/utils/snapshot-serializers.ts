function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function normalizeVersionOfPackage(str: string, pkg: string) {
  const regex = new RegExp(
    `("${escapeRegExp(pkg)}": "[~|^]\\d\\d?\\.)(\\d.\\d)"`,
  );
  return str.replace(regex, '$1X.X"');
}

const dependenciesToNormalize = [
  '@angular-devkit/build-angular',
  '@angular/cli',
  '@angular/compiler-cli',
  'ng-packagr',
  'typescript',
];

/**
 * Normalize dependencies controlled by the Angular CLI so that only the major version is explicitly
 * checked so that we can massively cut down on snapshot update noise across PRs.
 */
export const normalizeVersionsOfPackagesWeDoNotControl = {
  serialize(str: string) {
    for (const pkg of dependenciesToNormalize) {
      str = normalizeVersionOfPackage(str, pkg);
    }
    return str;
  },
  test(val: string) {
    // Only run if we think it's package.json contents
    return (
      val != null &&
      typeof val === 'string' &&
      dependenciesToNormalize.some((pkg) => val.includes(`"${pkg}"`))
    );
  },
};
