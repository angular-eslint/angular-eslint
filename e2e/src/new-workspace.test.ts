import path from 'node:path';
import { setWorkspaceRoot } from 'nx/src/utils/workspace-root';
import { FIXTURES_DIR, Fixture } from '../utils/fixtures';
import {
  LONG_TIMEOUT_MS,
  runNgAdd,
  runNgGenerate,
  runNgNew,
} from '../utils/local-registry-process';
import { runLint } from '../utils/run-lint';
import { normalizeVersionsOfPackagesWeDoNotControl } from '../utils/snapshot-serializers';

expect.addSnapshotSerializer(normalizeVersionsOfPackagesWeDoNotControl);

const fixtureDirectory = 'new-workspace';
let fixture: Fixture;

describe('new-workspace', () => {
  jest.setTimeout(LONG_TIMEOUT_MS);

  beforeEach(async () => {
    process.chdir(FIXTURES_DIR);
    await runNgNew(fixtureDirectory);

    process.env.NX_DAEMON = 'false';
    process.env.NX_CACHE_PROJECT_GRAPH = 'false';

    const workspaceRoot = path.join(FIXTURES_DIR, fixtureDirectory);
    process.chdir(workspaceRoot);
    process.env.NX_WORKSPACE_ROOT_PATH = workspaceRoot;
    setWorkspaceRoot(workspaceRoot);

    fixture = new Fixture(workspaceRoot);

    await runNgAdd();
    await runNgGenerate(['app', 'another-app', '--interactive=false']);
    await runNgGenerate(['lib', 'another-lib', '--interactive=false']);
  });

  it('it should pass linting after creating a new workspace from scratch using @angular-eslint', async () => {
    // TSLint configs and dependencies should not be present
    expect(fixture.fileExists('tslint.json')).toBe(false);
    expect(
      JSON.stringify(fixture.readJson('package.json').devDependencies, null, 2),
    ).toMatchSnapshot();

    // Root project
    expect(fixture.readFile('.eslintrc.json')).toMatchSnapshot();

    expect(
      fixture.readJson('angular.json').projects['new-workspace'].architect.lint,
    ).toMatchSnapshot();

    // Additional project ("another-app")
    expect(fixture.fileExists('projects/another-app/tslint.json')).toBe(false);
    expect(
      fixture.readFile('projects/another-app/.eslintrc.json'),
    ).toMatchSnapshot();

    expect(
      fixture.readJson('angular.json').projects['another-app'].architect.lint,
    ).toMatchSnapshot();

    // Additional library project ("another-lib")
    expect(fixture.fileExists('projects/another-lib/tslint.json')).toBe(false);
    expect(
      fixture.readFile('projects/another-lib/.eslintrc.json'),
    ).toMatchSnapshot();

    expect(
      fixture.readJson('angular.json').projects['another-lib'].architect.lint,
    ).toMatchSnapshot();

    const lintOutput = await runLint(fixtureDirectory);
    expect(lintOutput).toMatchSnapshot();
  });
});
