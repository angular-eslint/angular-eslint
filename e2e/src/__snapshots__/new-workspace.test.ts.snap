// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 1`] = `
{
  "@angular-devkit/build-angular": "^17.X.X",
  "@angular-eslint/builder": "0.0.0-e2e",
  "@angular-eslint/eslint-plugin": "0.0.0-e2e",
  "@angular-eslint/eslint-plugin-template": "0.0.0-e2e",
  "@angular-eslint/schematics": "0.0.0-e2e",
  "@angular-eslint/template-parser": "0.0.0-e2e",
  "@angular/cli": "^17.X.X",
  "@angular/compiler-cli": "^17.X.X",
  "@types/jasmine": "~5.1.0",
  "@typescript-eslint/eslint-plugin": "7.8.0",
  "@typescript-eslint/parser": "7.8.0",
  "eslint": "^8.57.0",
  "jasmine-core": "~5.1.0",
  "karma": "~6.4.0",
  "karma-chrome-launcher": "~3.2.0",
  "karma-coverage": "~2.2.0",
  "karma-jasmine": "~5.1.0",
  "karma-jasmine-html-reporter": "~2.1.0",
  "ng-packagr": "^17.X.X",
  "typescript": "~5.X.X"
}
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 2`] = `
"{
  \\"root\\": true,
  \\"ignorePatterns\\": [
    \\"projects/**/*\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"*.ts\\"
      ],
      \\"extends\\": [
        \\"eslint:recommended\\",
        \\"plugin:@typescript-eslint/recommended\\",
        \\"plugin:@angular-eslint/recommended\\",
        \\"plugin:@angular-eslint/template/process-inline-templates\\"
      ],
      \\"rules\\": {
        \\"@angular-eslint/directive-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"attribute\\",
            \\"prefix\\": \\"app\\",
            \\"style\\": \\"camelCase\\"
          }
        ],
        \\"@angular-eslint/component-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"element\\",
            \\"prefix\\": \\"app\\",
            \\"style\\": \\"kebab-case\\"
          }
        ]
      }
    },
    {
      \\"files\\": [
        \\"*.html\\"
      ],
      \\"extends\\": [
        \\"plugin:@angular-eslint/template/recommended\\",
        \\"plugin:@angular-eslint/template/accessibility\\"
      ],
      \\"rules\\": {}
    }
  ]
}
"
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 3`] = `
Object {
  "builder": "@angular-eslint/builder:lint",
  "options": Object {
    "lintFilePatterns": Array [
      "src/**/*.ts",
      "src/**/*.html",
    ],
  },
}
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 4`] = `
"{
  \\"extends\\": \\"../../.eslintrc.json\\",
  \\"ignorePatterns\\": [
    \\"!**/*\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"*.ts\\"
      ],
      \\"rules\\": {
        \\"@angular-eslint/directive-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"attribute\\",
            \\"prefix\\": \\"app\\",
            \\"style\\": \\"camelCase\\"
          }
        ],
        \\"@angular-eslint/component-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"element\\",
            \\"prefix\\": \\"app\\",
            \\"style\\": \\"kebab-case\\"
          }
        ]
      }
    },
    {
      \\"files\\": [
        \\"*.html\\"
      ],
      \\"rules\\": {}
    }
  ]
}
"
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 5`] = `
Object {
  "builder": "@angular-eslint/builder:lint",
  "options": Object {
    "lintFilePatterns": Array [
      "projects/another-app/**/*.ts",
      "projects/another-app/**/*.html",
    ],
  },
}
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 6`] = `
"{
  \\"extends\\": \\"../../.eslintrc.json\\",
  \\"ignorePatterns\\": [
    \\"!**/*\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"*.ts\\"
      ],
      \\"rules\\": {
        \\"@angular-eslint/directive-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"attribute\\",
            \\"prefix\\": \\"lib\\",
            \\"style\\": \\"camelCase\\"
          }
        ],
        \\"@angular-eslint/component-selector\\": [
          \\"error\\",
          {
            \\"type\\": \\"element\\",
            \\"prefix\\": \\"lib\\",
            \\"style\\": \\"kebab-case\\"
          }
        ]
      }
    },
    {
      \\"files\\": [
        \\"*.html\\"
      ],
      \\"rules\\": {}
    }
  ]
}
"
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 7`] = `
Object {
  "builder": "@angular-eslint/builder:lint",
  "options": Object {
    "lintFilePatterns": Array [
      "projects/another-lib/**/*.ts",
      "projects/another-lib/**/*.html",
    ],
  },
}
`;

exports[`new-workspace it should pass linting after creating a new workspace from scratch using @angular-eslint 8`] = `
"
Linting \\"new-workspace\\"...

All files pass linting.


Linting \\"another-app\\"...

All files pass linting.


Linting \\"another-lib\\"...

All files pass linting.
"
`;
